File: src/main/GO/leetcode/lists/offerII/Offer55.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/17 18:36
 */

type BSTIterator struct {
	arr []int
}

func ConstructorBSTIterator(root *TreeNode) (it BSTIterator) {
	it.inorder(root)
	return
}

func (it *BSTIterator) inorder(node *TreeNode) {
	if node == nil {
		return
	}
	it.inorder(node.Left)
	it.arr = append(it.arr, node.Val)
	it.inorder(node.Right)
}

func (it *BSTIterator) Next() int {
	val := it.arr[0]
	it.arr = it.arr[1:]
	return val
}

func (it *BSTIterator) HasNext() bool {
	return len(it.arr) > 0
}

File: src/main/java/leetcode/lists/offerII/Offer55.java
package leetcode.lists.offerII;

import common.TreeNode;

import java.util.Deque;
import java.util.LinkedList;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/7/25 17:11
 */
public class Offer55 {

    class BSTIterator {

        private final Deque<TreeNode> stack;
        private TreeNode cur;

        public BSTIterator(TreeNode root) {
            cur = root;
            stack = new LinkedList<>();
        }

        public int next() {
            while (cur != null) {
                stack.push(cur);
                cur = cur.left;
            }
            cur = stack.pop();
            int ret = cur.val;
            cur = cur.right;
            return ret;
        }

        public boolean hasNext() {
            return cur != null || !stack.isEmpty();
        }
    }
}
